testing/base/source_gen.cpp

```diff
@@ -22,6 +22,237 @@ auto SourceGen::Global() -> SourceGen& {
+        << " candidates, an initial index of " << initial_last_type_name_index
+        << ", and with " << class_names_.size() << " classes left to emit!";
+  }
+}
+
+// We want to synthesize shuffled type names from the class names that will be
+// declared, but mixing in `i32` and other builtin types. A somewhat hack-y
+// approach, but currently the strategy is to map every type shorter than the
+// `MinClassNameLength` to one of our builtin types and the longer names to
+// class names in the state.
+//
+// We use non-unique random identifiers to get a sequence of lengths, and map
+// each identifier to one of our desired types as above. This allows repetition
+// as the identifiers repeat. We can also create a larger pool of builtin type
+// spellings eventually and use the different short identifier spellings to
+// select from this pool.
+//
+// When searching for a given length, we allow wrapping around the class names.
+// This may map several random identifiers to the same declared class name.
+//
+// TODO: Variation in how many times a name is repeated may introduce some
+// unnecessary noise, we should consider a fixed distribution of duplication
+// rather than just accepting whatever occurs randomly. Currently this only
+// stabilizes the total text size resulting from referencing prior type names.
+auto SourceGen::ClassGenState::BuildClassAndTypeNames(
+    SourceGen& gen, int num_classes, int num_types,
+    const TypeUseParams& type_use_params) -> void {
+  // Initially get the sequence of class names without shuffling so we can
+  // compute our type name pool from them prior to any shuffling.
+  class_names_ =
+      gen.GetUniqueIdentifiers(num_classes, /*min_length=*/MinClassNameLength);
+
+  type_names_.reserve(num_types);
+
+  // Compute the sum of weights and pre-process the fixed types.
+  int type_weight_sum = type_use_params.declared_types_weight;
+  for (const auto& fixed_type_weight : type_use_params.fixed_type_weights) {
+    type_weight_sum += fixed_type_weight.weight;
+    // Add all the fixed type spellings as immediately valid.
+    valid_type_names_.Insert(gen.IsCpp() ? fixed_type_weight.cpp_spelling
+                                         : fixed_type_weight.carbon_spelling);
+  }
+
+  // Compute the number of declared types used. We expect to have a decent
+  // number of repeated names, so we repeatedly append the entire sequence of
+  // class names until there is some remainder of names needed.
+  int num_declared_types =
+      num_types * type_use_params.declared_types_weight / type_weight_sum;
+  for ([[maybe_unused]] auto _ : llvm::seq(num_declared_types / num_classes)) {
+    type_names_.append(class_names_.begin(), class_names_.end());
+  }
+  // Now append the remainder number of class names. This is where the class
+  // names being un-shuffled is essential. We're going to have one extra
+  // reference to some fraction of the class names and we want that to be a
+  // stable fraction.
```

Comment:
This change is to emphasize it is the names that are stable, not the number represented by the fraction.

```suggestion
  // stable subset.
```